src/Qir/Runtime/lib/QSharpFoundation/qsharp__foundation__qis.hpp

```diff
@@ -11,6 +11,16 @@
 struct QirArray;
 struct QirCallable;
 
+struct QirAssertMeasurementProbabilityTuple
+{
+    QirArray* bases;
+    QirArray* qubits;
+    RESULT* result;
+    double prob;
+    QirString* msg;
+    double tol;
+};
```

Comment:
(Nit: In "extern C" section, if that is applicable to structs)

Reply:
See my comment above. I think either all these capture tuple structs should be extern C together or all should be C++ structs like `QirArray` but I'm not confident which approach makes sense. They aren't exactly like the opaque pointers, but they also aren't fully a part of the API. Maybe I should just forward declare them here and then move the definitions into the cpp file?

Reply:
"Forward declaring them here and moving the definitions into the cpp file" seems to me a quite reasonable future alternative to the item [Avoid Exposing Entities With Multiple Layouts](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/QIR-RT-API-Design-Guidelines.md#avoid-exposing-entities-with-multiple-layouts). However, if we expose externally just the forward declaration, then we will have to expose also the functions that 
* are `new`-like (allocate and initialize the instance),
* manipulate the instance (getters and setters),
* are `delete`-like (deinitialize and deallocate the instance)

(feel free to leave as is in this PR).

Reply:
You are totally right. I think we are in a tricky spot with the `%Tuple` type, because it straddles the boundary between struct and opaque type. The methods for creation, manipulation, and deletion (via reference counting), as described here: https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Data-Types.md#tuples-and-user-defined-types. However, unlike the opaque types we use for arrays, the returned memory is accessed directly by the client, making it "semi-opaque" which is odd. 

I think this might be another place where we need to consider filing feedback against the QIR spec and figuring out the right design to resolve this ambiguity. Let's sync up more on this sometime this week, as I could really use your insight on this one!

Reply:
Sure, let's sync up.